File: scala/anagrams.scala
def swap(a: Array[Char], i: Int, j: Int): Unit =
  val saved = a(i)
  a(i) = a(j)
  a(j) = saved

def generatePermutations(a: Array[Char], n: Int): Unit =
  if n == 0 then
    println(String.valueOf(a))
  else
    for i <- 0 until n do
      generatePermutations(a, n-1)
      swap(a, if n % 2 == 0 then 0 else i, n)
    generatePermutations(a, n-1)

@main def anagrams(args: String*) =
  if args.length != 1 then
    Console.err.println("Exactly one argument is required")
    sys.exit(1)
  var word = args(0)
  generatePermutations(word.toCharArray, word.length-1)

File: clojure/anagrams.clj
(defn swap [a i j]
  (assoc a j (a i) i (a j)))

(defn generatePermutations [v n]
  (if (zero? n)
    (do (println (apply str v)) v)

    (loop [i 0 a v]
      (if (< i n)
        (do
          (let [a (generatePermutations a (dec n))]
          (recur (inc i) (swap a (if (even? n) 0 i) n))))
        (generatePermutations a (dec n))))))

(if (not= (count *command-line-args*) 1)
  (do
    (println "Exactly one argument is required")
    (System/exit 1))
  (let [word (->> *command-line-args* first vec)]
    (generatePermutations word (dec (count word)))))
